state-chain/pallets/cf-elections/src/tests.rs

```diff
@@ -1,35 +1,138 @@
+				"Expected one shared data hash."
+			);
+		})
+		// Delete the election when we finalize: should cause all refs to be deleted too.
+		.update_settings(&[BehaviourUpdate::DeleteOnFinalizeConsensus(true)])
+		.snapshot();
+
+	// Case 1: Provide Shared Data through provide_shared_data extrinsic:
+	let case_1 = TestRunner::from_snapshot(initial_test_state.clone())
+		.assert_calls_ok(&authorities[..1], |_| Call::<Test, Instance1>::provide_shared_data {
+			shared_data: (),
+		});
+
+	// Case 2: Provide Shared Data through vote extrinsic:
+	let case_2 = TestRunner::from_snapshot(initial_test_state.clone()).submit_votes(
+		&authorities[..1],
+		AuthorityVote::Vote(()),
+		None,
+	);
+
+	for (label, test_case) in [(1, case_1), (2, case_2)] {
+		test_case
+			// Shared data provided, all votes should now be counted.
+			.expect_consensus(ConsensusStatus::Changed {
+				previous: 0,
+				new: authorities.len() as AuthorityCount,
+			})
+			.then_execute_with_keep_context(|_| {
+				assert!(
+					SharedDataReferenceCount::<Test, _>::iter().next().is_none(),
+					"Case {label}: Expected shared data refs to be removed but found: {:?}, components are: {:?}",
+					SharedDataReferenceCount::<Test, _>::iter().collect::<Vec<_>>(),
+					IndividualComponents::<Test, _>::iter().collect::<Vec<_>>(),
+				);
+			});
+	}
+}
+
+#[test]
+fn ensure_can_vote() {
+	new_test_ext().then_execute_at_next_block(|()| {
+		let setup = TestSetup { num_non_contributing_authorities: 1, ..Default::default() };
+
+		let initial_state = election_test_ext(setup.clone())
+			.new_election()
+			.submit_votes(
+				&setup.non_contributing_authorities()[..],
+				AuthorityVote::Vote(()),
+				Some(Error::NotContributing),
+			)
+			.snapshot();
+
+		// Contributing authorities can vote.
+		TestRunner::from_snapshot(initial_state.clone()).submit_votes(
+			&setup.contributing_authorities()[..],
+			AuthorityVote::Vote(()),
+			None,
+		);
+
+		// If governance pauses elections, no votes can be submitted.
+		TestRunner::from_snapshot(initial_state.clone())
+			.then_apply_extrinsics(|_| {
+				[(OriginTrait::root(), Call::<Test, _>::pause_elections {}, Ok(()))]
+			})
+			.submit_votes(
+				&setup.all_authorities()[..],
+				AuthorityVote::Vote(()),
+				Some(Error::Paused),
+			);
+	});
+}
+
 pub trait ElectoralSystemTestExt: Sized {
+	fn update_settings(self, updates: &[BehaviourUpdate]) -> Self;
+	fn expect_consensus_after_next_block(self, expected: ConsensusStatus<AuthorityCount>) -> Self;
 	fn assume_consensus(self) -> Self;
 	fn assume_no_consensus(self) -> Self;
 	fn expect_consensus(self, expected: ConsensusStatus<AuthorityCount>) -> Self;
 	fn new_election(self) -> Self;
-	fn submit_simple<I: 'static>(
-		self,
-		validator_ids: &[u64],
-		call: impl Fn(&u64) -> Call<Test, I>,
-	) -> Self
-	where
-		Test: Config<I>;
 	fn submit_votes<I: 'static>(
 		self,
 		validator_ids: &[u64],
 		vote: AuthorityVoteOf<MockElectoralSystem>,
 		expected_error: Option<Error<Test, I>>,
```

Comment:
does feel a little strange to use Option here - why not expect Result<(), Error> .... we end up converting it to result anyway.

Reply:
Yeah I think I agree - the idea was that we usually don't expect an error so we could just use Default::default, but agree it's bit unexpected..